react/features/mobile/callkit/CallKit.js

```diff
@@ -0,0 +1,235 @@
+import {
+    NativeModules,
+    NativeEventEmitter,
+    Platform
+} from 'react-native';
+
+const RNCallKit = NativeModules.RNCallKit;
+
+/**
+ * Thin wrapper around Apple's CallKit functionality.
+ *
+ * In CallKit requests are performed via actions (either user or system started)
+ * and async events are reported via dedicated methods. This class exposes that
+ * functionality in the form of methods and events. One important thing to note
+ * is that even if an action is started by the system (because the user pressed
+ * the "end call" button in the CallKit view, for example) the event will be
+ * emitted in the same way as it would if the action originated from calling
+ * the "endCall" method in this class, for example.
+ *
+ * Emitted events:
+ *  - performAnswerCallAction: The user pressed the answer button.
+ *  - performEndCallAction: The call should be ended.
+ *  - performSetMutedCallAction: The call muted state should change. The
+ *    ancillary `data` object contains a `muted` attribute.
+ *  - providerDidReset: The system has reset, all calls should be terminated.
+ *    This event gets no associated data.
+ *
+ * All events get a `data` object with a `callUUID` property, unless stated
+ * otherwise.
+ */
+class CallKit extends NativeEventEmitter {
+    /**
+     * Initializes a new {@code CallKit} instance.
+     */
+    constructor() {
+        super(RNCallKit);
+        this._setup = false;
+    }
+
+    /**
+     * Returns True if the current platform is supported, false otherwise. The
+     * supported platforms are: iOS >= 10.
+     *
+     * @private
+     * @returns {boolean}
+     */
+    static isSupported() {
+        return Platform.OS === 'ios' && parseInt(Platform.Version, 10) >= 10;
+    }
+
+    /**
+     * Checks if CallKit was setup, and throws an exception in that case.
+     *
+     * @private
+     * @returns {void}
+     */
+    _checkSetup() {
+        if (!this._setup) {
+            throw new Error('CallKit not initialized, call setup() first.');
+        }
+    }
```

Comment:
I found the `CallKit` implementation and the way it was used very complex. I modified `RNCallKit` to not load unless CallKit is supported, removed the need for `_checkSetup` and `isSupported`, and reduced it to adding an `addListener` method. I don't like complexity and repetition (the two support checks in Objective-C and JavaScript) because it makes maintenance harder (for me).